File: dotnet/src/xjustiz.core-dotnet/Models/Entities/Core/HashInfoCore.cs
namespace xjustiz.core_dotnet.Models.Entities.Core;

using System.Text.Json.Serialization;
using System.Xml.Serialization;
using xjustiz.core_dotnet.Models.Helpers;
using xjustiz.core_dotnet.Util.Versioning;

/// <summary>
/// Repräsentiert Integritätsinformationen einer Datei.<br/>
/// <u><b>Hash info:</b></u> Represents file integrity information.
/// </summary>
[XJustizCoreAvailability(XJustizCoreVersion.V0_2_0)]
public class HashInfoCore : IEquatable<HashInfoCore>
{
    /// <summary>
    /// Der verwendete Hash-Algorithmus (z.B. SHA-256).<br/>
    /// <u><b>Algorithm:</b></u> The hash algorithm used (e.g., SHA-256).
    /// </summary>
    [XmlElement("algorithmus", Namespace = SchemeConstants.XJustizCore_Tns)]
    [JsonPropertyName("algorithm")]
    [XJustizCoreAvailability(XJustizCoreVersion.V0_2_0)]
    public string Algorithm { get; set; } = "SHA-256";

    /// <summary>
    /// Der Hashwert.<br/>
    /// <u><b>Value:</b></u> The hash value.
    /// </summary>
    [XmlElement("wert", Namespace = SchemeConstants.XJustizCore_Tns)]
    [JsonPropertyName("value")]
    [XJustizCoreAvailability(XJustizCoreVersion.V0_2_0)]
    public string Value { get; set; } = string.Empty;

    public static bool operator ==(HashInfoCore? left, HashInfoCore? right) => Equals(left, right);

    public static bool operator !=(HashInfoCore? left, HashInfoCore? right) => !Equals(left, right);

    public bool Equals(HashInfoCore? other)
    {
        if (other is null)
        {
            return false;
        }

        if (ReferenceEquals(this, other))
        {
            return true;
        }

        // Use OrdinalIgnoreCase for Algorithm as naming conventions can vary,
        // but hashes are usually compared with Ordinal.
        return string.Equals(Algorithm, other.Algorithm, StringComparison.OrdinalIgnoreCase) &&
               string.Equals(Value, other.Value, StringComparison.Ordinal);
    }

    public override bool Equals(object? obj)
    {
        return Equals(obj as HashInfoCore);
    }

    public override int GetHashCode()
    {
        // HashCode.Combine is the modern, robust way to generate a hash code
        return HashCode.Combine(
            Algorithm?.ToLowerInvariant(),
            Value);
    }
}

File: sdk/java/src/main/java/de/xjustiz/core/models/HashInfoCore.java
package de.xjustiz.core.models;

import com.fasterxml.jackson.annotation.JsonAlias;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlProperty;
import jakarta.xml.bind.annotation.XmlAccessType;
import jakarta.xml.bind.annotation.XmlAccessorType;
import jakarta.xml.bind.annotation.XmlElement;
import jakarta.xml.bind.annotation.XmlType;

/**
 * Repräsentiert Integritätsinformationen einer Datei.
 * <p>
 * <u><b>Hash info:</b></u> Represents file integrity information.
 */
@XmlType(name = "HashInfoCore", propOrder = { "algorithm", "value" })
@XmlAccessorType(XmlAccessType.FIELD)
@JsonIgnoreProperties(ignoreUnknown = true)
@JsonInclude(JsonInclude.Include.NON_EMPTY)
public class HashInfoCore {

    /**
     * Der verwendete Hash-Algorithmus (z.B. SHA-256).
     * <p>
     * <u><b>Algorithm:</b></u> The hash algorithm used (e.g., SHA-256).
     */
    @XmlElement(name = "algorithmus", namespace = "http://www.xjustizcore.de")
    @JacksonXmlProperty(localName = "algorithmus", namespace = "http://www.xjustizcore.de")
    @JsonProperty("algorithm")
    @JsonAlias({ "Algorithm", "algorithm" })
    private String algorithm;

    /**
     * Der Hashwert.
     * <p>
     * <u><b>Value:</b></u> The hash value.
     */
    @XmlElement(name = "wert", namespace = "http://www.xjustizcore.de")
    @JacksonXmlProperty(localName = "wert", namespace = "http://www.xjustizcore.de")
    @JsonProperty("value")
    @JsonAlias({ "Value", "value" })
    private String value;

    /**
     * Default constructor.
     */
    public HashInfoCore() {}

    /**
     * Gets the value of the Algorithm property.
     * @return the value of the Algorithm property
     */
    public String getAlgorithm() { return algorithm; }
    /**
     * Sets the value of the Algorithm property.
     * @param algorithm the value to set
     */
    public void setAlgorithm(String algorithm) { this.algorithm = algorithm; }

    /**
     * Gets the value of the Value property.
     * @return the value of the Value property
     */
    public String getValue() { return value; }
    /**
     * Sets the value of the Value property.
     * @param value the value to set
     */
    public void setValue(String value) { this.value = value; }

}